Fix dialog losing focus when changing WindowState

corvinsz opened the pull request:
fixes #3434 

The fix is basically to listen for the `StateChanged` event of the parent `Window`.
If the window is restored from the minimized state, we focus the previously focused element, which got stored in a field `_lastFocusedDialogElement` when the window got minimized.

I want to point out the artificial delay I added, that we have in other places in the DialogHost class too. I just couldn't get it to work without the (50ms) delay, as described in my comment above the "hacky" code.
```csharp
...
// Kinda hacky, but without a delay the focus doesn't always get set correctly because the Focus() method fires too early
Task.Delay(50).ContinueWith(_ => this.Dispatcher.BeginInvoke(DispatcherPriority.Background, new Action(() =>
...
```

<!-- gk-ai-analysis-start:34082b4928b5955fcd150b7f51f2985072b8bd83 -->
<!-- gk-ai-analysis-data:eyJzdW1tYXJ5IjoiQWRkcyBhIFVJIHRlc3Qgc2FtcGxlIHdpdGggbXVsdGlwbGUgZm9jdXNhYmxlIGVsZW1lbnRzIHRvIGZhY2lsaXRhdGUgdGVzdGluZyBvZiBEaWFsb2dIb3N0IGZvY3VzIHJlc3RvcmF0aW9uIGxvZ2ljIGR1cmluZyB3aW5kb3cgc3RhdGUgdHJhbnNpdGlvbnMuIiwia2V5SW5zaWdodHMiOlt7InRpdGxlIjoiRm9jdXMgdmFsaWRhdGlvbiB0ZXN0IHNhbXBsZSIsImRlc2NyaXB0aW9uIjoiSW50cm9kdWNlcyBhIGRlZGljYXRlZCAnV2l0aE11bHRpcGxlVGV4dEJveGVzJyBVc2VyQ29udHJvbCB0byB0aGUgVUkgdGVzdCBwcm9qZWN0LCBwcm92aWRpbmcgYSByZXByb2R1Y2libGUgZW52aXJvbm1lbnQgd2l0aCBzZXZlcmFsIGlucHV0IGZpZWxkcyB0byB2ZXJpZnkgZm9jdXMgcGVyc2lzdGVuY2UuIiwic2V2ZXJpdHkiOiJsb3ciLCJmaWxlUGF0aCI6InRlc3RzL01hdGVyaWFsRGVzaWduVGhlbWVzLlVJVGVzdHMvU2FtcGxlcy9EaWFsb2dIb3N0L1dpdGhNdWx0aXBsZVRleHRCb3hlcy54YW1sIiwibGluZVN0YXJ0IjoxLCJsaW5lRW5kIjoyOH0seyJ0aXRsZSI6IkF1dG9tYXRlZCBkaWFsb2cgYWN0aXZhdGlvbiBmb3IgdGVzdGluZyIsImRlc2NyaXB0aW9uIjoiVGhlIHRlc3Qgc2FtcGxlIGF1dG9tYXRpY2FsbHkgb3BlbnMgdGhlIERpYWxvZ0hvc3QgdXBvbiBsb2FkaW5nIHZpYSB0aGUgRGlhbG9nSG9zdF9Mb2FkZWQgZXZlbnQgaGFuZGxlciwgZW5zdXJpbmcgdGhlIGRpYWxvZyBpcyBhY3RpdmUgZm9yIFVJIGF1dG9tYXRpb24gd2l0aG91dCBtYW51YWwgc3RlcHMuIiwic2V2ZXJpdHkiOiJsb3ciLCJmaWxlUGF0aCI6InRlc3RzL01hdGVyaWFsRGVzaWduVGhlbWVzLlVJVGVzdHMvU2FtcGxlcy9EaWFsb2dIb3N0L1dpdGhNdWx0aXBsZVRleHRCb3hlcy54YW1sLmNzIiwibGluZVN0YXJ0IjoxMiwibGluZUVuZCI6MTV9LHsidGl0bGUiOiJNdWx0aS10YXJnZXQgZm9jdXMgbGF5b3V0IiwiZGVzY3JpcHRpb24iOiJJbmNsdWRlcyB0aHJlZSBkaXN0aW5jdCBUZXh0Qm94IGNvbnRyb2xzIHdpdGhpbiB0aGUgZGlhbG9nIGNvbnRlbnQgdG8gc3BlY2lmaWNhbGx5IHRlc3QgaWYgZm9jdXMgcmV0dXJucyB0byB0aGUgY29ycmVjdCBlbGVtZW50IGFtb25nIG11bHRpcGxlIHBvdGVudGlhbCB0YXJnZXRzLiIsInNldmVyaXR5IjoibG93IiwiZmlsZVBhdGgiOiJ0ZXN0cy9NYXRlcmlhbERlc2lnblRoZW1lcy5VSVRlc3RzL1NhbXBsZXMvRGlhbG9nSG9zdC9XaXRoTXVsdGlwbGVUZXh0Qm94ZXMueGFtbCIsImxpbmVTdGFydCI6MTYsImxpbmVFbmQiOjI0fV0sImlzc3VlcyI6W10sInN1Z2dlc3Rpb25zIjpbeyJ0aXRsZSI6IkRlZmluZSBpbml0aWFsIGZvY3VzIGZvciBwcmVkaWN0YWJpbGl0eSIsImRlc2NyaXB0aW9uIjoiQ29uc2lkZXIgc2V0dGluZyB0aGUgSW5pdGlhbEZvY3VzRWxlbWVudCBwcm9wZXJ0eSBvbiB0aGUgRGlhbG9nSG9zdCB0byBvbmUgb2YgdGhlIFRleHRCb3hlcyB0byBlbnN1cmUgYSBkZXRlcm1pbmlzdGljIHN0YXJ0aW5nIHN0YXRlIGZvciB0aGUgZm9jdXMtcmVsYXRlZCBVSSB0ZXN0cy4iLCJzZXZlcml0eSI6ImxvdyIsImZpbGVQYXRoIjoidGVzdHMvTWF0ZXJpYWxEZXNpZ25UaGVtZXMuVUlUZXN0cy9TYW1wbGVzL0RpYWxvZ0hvc3QvV2l0aE11bHRpcGxlVGV4dEJveGVzLnhhbWwiLCJsaW5lU3RhcnQiOjExfSx7InRpdGxlIjoiVXNlIERpc3BhdGNoZXIgZm9yIGRpYWxvZyBvcGVuaW5nIiwiZGVzY3JpcHRpb24iOiJPcGVuaW5nIGEgZGlhbG9nIGltbWVkaWF0ZWx5IHdpdGhpbiB0aGUgTG9hZGVkIGV2ZW50IGNhbiBvY2Nhc2lvbmFsbHkgbGVhZCB0byB0aW1pbmcgaXNzdWVzIGluIFdQRi4gV3JhcHBpbmcgdGhlIElzT3BlbiBhc3NpZ25tZW50IGluIGEgRGlzcGF0Y2hlci5CZWdpbkludm9rZSBjYWxsIG1heSBpbmNyZWFzZSB0ZXN0IHJlbGlhYmlsaXR5LiIsInNldmVyaXR5IjoibG93IiwiZmlsZVBhdGgiOiJ0ZXN0cy9NYXRlcmlhbERlc2lnblRoZW1lcy5VSVRlc3RzL1NhbXBsZXMvRGlhbG9nSG9zdC9XaXRoTXVsdGlwbGVUZXh0Qm94ZXMueGFtbC5jcyIsImxpbmVTdGFydCI6MTR9XSwic2VjdXJpdHkiOltdfQ== -->
<!-- gk-ai-analysis-end:34082b4928b5955fcd150b7f51f2985072b8bd83 -->

<!-- gitkraken-review-badge-begin -->
---
<a href="https://gitkraken.dev/review/github/MaterialDesignInXAML/MaterialDesignInXamlToolkit/pull/3923">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://gitkraken.dev/images/figures/gitkraken-review-badge-dark.svg">
    <img src="https://gitkraken.dev/images/figures/gitkraken-review-badge-light.svg" alt="Open with GitKraken">
  </picture>
</a>
<!-- gitkraken-review-badge-end -->

corvinsz committed:
restore focus to the last focused element inside of the dialog

corvinsz committed:
add tests

corvinsz committed:
undo whitespace changes

corvinsz commented:
It seems like there is a problem with running the tests via github actions. This PR just had another failing test (`SplitButton_WithButtonInPopup_CanBeInvoked`). When reran, the test passed.
I'm not sure if this is a TUnit or github actions problem?

I had the same issue in #3915

Keboo commented on src/MaterialDesignThemes.Wpf/DialogHost.cs:
```diff
@@ -405,6 +409,50 @@ private static void IsOpenPropertyChangedCallback(DependencyObject dependencyObj
+    
+    private void ListenForWindowStateChanged(Window? window)
+    {
+        window ??= Window.GetWindow(this);
```
Not sure we need the `??=` as the only caller used this same call to try and get the window, so if that is null, this will very likely return null a second time.
```suggestion
```

Keboo commented on src/MaterialDesignThemes.Wpf/DialogHost.cs:
```diff
@@ -405,6 +409,50 @@ private static void IsOpenPropertyChangedCallback(DependencyObject dependencyObj
+
+        if (window is not null)
+        {
+            window.StateChanged += Window_StateChanged;
```
We want to make sure we don't register twice.
```suggestion
            window.StateChanged -= Window_StateChanged;
            window.StateChanged += Window_StateChanged;
```

Keboo commented on src/MaterialDesignThemes.Wpf/DialogHost.cs:
```diff
@@ -405,6 +409,50 @@ private static void IsOpenPropertyChangedCallback(DependencyObject dependencyObj
+
+        // We only need to focus anything manually if the window changes state from Minimized --> (Normal or Maximized)
+        // Going from Normal --> Maximized (and vice versa) is fine since the focus is already kept correctly
+        if (IsWindowRestoredFromMinimized() && IsLastFocusedDialogElementFocusable())
+        {
+            // Kinda hacky, but without a delay the focus doesn't always get set correctly because the Focus() method fires too early
+            Task.Delay(50).ContinueWith(_ => this.Dispatcher.BeginInvoke(DispatcherPriority.Background, new Action(() =>
+            {
+                _lastFocusedDialogElement!.Focus();
+            })));
+        }
```
Because Focusable and Visible can change I would move this check inside of the Task.Delay.


```suggestion
        if (IsWindowRestoredFromMinimized())
        {
            // Kinda hacky, but without a delay the focus doesn't always get set correctly because the Focus() method fires too early
            Task.Delay(50).ContinueWith(_ => this.Dispatcher.BeginInvoke(DispatcherPriority.Background, new Action(() =>
            {
                if (IsLastFocusedDialogElementFocusable())
                {
                    _lastFocusedDialogElement!.Focus();
                }
            })));
        }
```

corvinsz committed:
Apply suggestions from code review

LGTM, thank you for the review!

Co-authored-by: Kevin B <Keboo@users.noreply.github.com>

corvinsz committed:
Adjust test to test against both DialogHost styles

corvinsz reviewed (commented):
Discussion on how to get the last focused element

corvinsz commented:
I also adjusted the test to test against both DialogHost styles. I would say this problem is worth testing for both styles.

corvinsz committed:
Refactor DialogHost to use HwndSourceHook for focus

Replaced the Window.StateChanged event-based approach with an
HwndSourceHook-based implementation to handle window state changes
(minimize/restore) more effectively. Removed `_previousWindowState`
and introduced `_hook` for managing system commands (`WM_SYSCOMMAND`).

Added a `Hook` method to handle `SC_MINIMIZE` and `SC_RESTORE`
commands, ensuring proper focus management. Refactored focus
restoration logic to use asynchronous behavior (`Task.Delay` and
`Dispatcher.BeginInvoke`) for improved reliability. Removed redundant
state-checking methods and improved code clarity with constants for
system commands.
